thrimbletrimmer/scripts/common.js

```diff
@@ -444,22 +418,192 @@ function dateTimeFromVideoPlayerTime(videoPlayerTime) {
+	for (const segment of segmentList) {
+		const segmentStart = DateTime.fromISO(segment.rawProgramDateTime);
+		const segmentEnd = segmentStart.plus({ seconds: segment.duration });
+		if (dateTime >= segmentStart && dateTime <= segmentEnd) {
+			return segment.start + dateTime.diff(segmentStart).as("seconds");
+		}
+	}
+	return null;
+}
+
+function videoHumanTimeFromDateTime(dateTime) {
+	const videoPlayerTime = videoPlayerTimeFromDateTime(dateTime);
+	if (videoPlayerTime === null) {
+		return null;
+	}
+	return videoHumanTimeFromVideoPlayerTime(videoPlayerTime);
+}
+
+function assembleVideoPlaylistURL(basePlaylistURL) {
+	let playlistURL = basePlaylistURL;
+
+	const queryStringParts = startAndEndTimeQueryStringParts();
+	if (queryStringParts) {
+		playlistURL += "?" + queryStringParts.join("&");
+	}
+	return playlistURL;
+}
+
+function startAndEndTimeQueryStringParts() {
+	const startTime = getStartTime();
+	const endTime = getEndTime();
+
+	let queryStringParts = [];
+	if (startTime) {
+		queryStringParts.push(`start=${wubloaderTimeFromDateTime(startTime)}`);
+	}
+	if (endTime) {
+		queryStringParts.push(`end=${wubloaderTimeFromDateTime(endTime)}`);
+	}
+	return queryStringParts;
+}
+
+function getSegmentList() {
+	return globalPlayer.latencyController.levelDetails.fragments;
+}
+
+function hasSegmentList() {
+	if (
+		globalPlayer &&
+		globalPlayer.latencyController &&
+		globalPlayer.latencyController.levelDetails &&
+		globalPlayer.latencyController.levelDetails.fragments
+	) {
+		return true;
+	}
+	return false;
+}
+
 function downloadFrame() {
 	const videoElement = document.getElementById("video");
 	const dateTime = dateTimeFromVideoPlayerTime(videoElement.currentTime);
-	const url = `/frame/${globalStreamName}/source.png?timestamp=${wubloaderTimeFromDateTime(dateTime)}`;
+	const url = `/frame/${globalStreamName}/source.png?timestamp=${wubloaderTimeFromDateTime(
+		dateTime
+	)}`;
 	// Avoid : as it causes problems on Windows
 	const filename = `${dateTime.toFormat("yyyy-LL-dd'T'HH-mm-ss.SSS")}.png`;
 	triggerDownload(url, filename);
 }
 
 function triggerDownload(url, filename) {
 	// URL must be same-origin.
-	const link = document.createElement('a');
-	link.setAttribute('download', filename);
+	const link = document.createElement("a");
+	link.setAttribute("download", filename);
 	link.href = url;
-	link.setAttribute('target', '_blank');
+	link.setAttribute("target", "_blank");
 	document.body.appendChild(link);
 	link.click();
 	document.body.removeChild(link);
 }
+
+async function getChatLog(startWubloaderTime, endWubloaderTime) {
+	globalChatData = [];
+	const url = `/${globalStreamName}/chat.json?start=${startWubloaderTime}&end=${endWubloaderTime}`;
+	const response = await fetch(url);
+	if (!response.ok) {
+		return;
+	}
+	const chatLogData = await response.json();
+	for (const chatLine of chatLogData) {
+		if (
+			chatLine.command !== "PRIVMSG" &&
+			chatLine.command !== "CLEARMSG" &&
+			chatLine.command !== "CLEARCHAT"
```

Comment:
You might also want to support USERNOTICE (displaying the text in `chatLine.tags["system-msg"]`). This includes sub messages (admittedly not a big deal for us) and raids.

Reply:
Ah, right. This didn't really come in my test data, but I'll see if I can add it at some point.